Incremental Template Type-Checking

alxhub opened the pull request:
# Take 2 of incremental template type-checking

This PR comprises 4 commits which are best reviewed separately.

alxhub committed:
fix(compiler): switch to 'referencedFiles' for shim generation

Shim generation was built on a lie.

Shims are files added to the program which aren't original files authored by
the user, but files authored effectively by the compiler. These fall into
two categories: files which will be generated (like the .ngfactory shims we
generate for View Engine compatibility) as well as files used internally in
compilation (like the __ng_typecheck__.ts file).

Previously, shim generation was driven by the `rootFiles` passed to the
compiler as input. These are effectively the `files` listed in the
`tsconfig.json`. Each shim generator (e.g. the `FactoryGenerator`) would
examine the `rootFiles` and produce a list of shim file names which it would
be responsible for generating. These names would then be added to the
`rootFiles` when the program was created.

The fatal flaw here is that `rootFiles` does not always account for all of
the files in the program. In fact, it's quite rare that it does. Users don't
typically specify every file directly in `files`. Instead, they rely on
TypeScript, during program creation, starting with a few root files and
transitively discovering all of the files in the program.

This happens, however, during `ts.createProgram`, which is too late to add
new files to the `rootFiles` list.

As a result, shim generation was only including shims for files actually
listed in the `tsconfig.json` file, and not for the transitive set of files
in the user's program as it should.

This commit completely rewrites shim generation to use a different technique
for adding files to the program, inspired by View Engine's shim generator.
In this new technique, as the program is being created and `ts.SourceFile`s
are being requested from the `NgCompilerHost`, shims for those files are
generated and a reference to them is patched onto the original file's
`ts.SourceFile.referencedFiles`. This causes TS to think that the original
file references the shim, and causes the shim to be included in the program.
The original `referencedFiles` array is saved and restored after program
creation, hiding this little hack from the rest of the system.

The new shim generation engine differentiates between two kinds of shims:
top-level shims (such as the flat module entrypoint file and
__ng_typecheck__.ts) and per-file shims such as ngfactory or ngsummary
files. The former are included via `rootFiles` as before, the latter are
included via the `referencedFiles` of their corresponding original files.

As a result of this change, shims are now correctly generated for all files
in the program, not just the ones named in `tsconfig.json`.

A few mitigating factors prevented this bug from being realized until now:

* in g3, `files` does include the transitive closure of files in the program
* in CLI apps, shims are not really used

This change also makes use of a novel technique for associating information
with source files: the use of an `NgExtension` `Symbol` to patch the
information directly onto the AST object. This is used in several
circumstances:

* For shims, metadata about a `ts.SourceFile`'s status as a shim and its
  origins are held in the extension data.
* For original files, the original `referencedFiles` are stashed in the
  extension data for later restoration.

The main benefit of this technique is a lot less bookkeeping around `Map`s
of `ts.SourceFile`s to various kinds of data, which need to be tracked/
invalidated as part of incremental builds.

This technique is based on designs used internally in the TypeScript
compiler and is serving as a prototype of this design in ngtsc. If it works
well, it could have benefits across the rest of the compiler.

alxhub committed:
perf(compiler-cli): split Ivy template type-checking into multiple files

As a performance optimization, this commit splits the single
__ngtypecheck__.ts file which was previously added to the user's program as
a container for all template type-checking code into multiple .ngtypecheck
shim files, one for each original file in the user's program.

In larger applications, the generation, parsing, and checking of this single
type-checking file was a huge performance bottleneck, with the file often
exceeding 1 MB in text content. Particularly in incremental builds,
regenerating this single file for the entire application proved especially
expensive.

This commit introduces a new strategy for template type-checking code which
makes use of a new interface, the `TypeCheckingProgramStrategy`. This
interface abstracts the process of creating a new `ts.Program` to type-check
a particular compilation, and allows the mechanism there to be kept separate
from the more complex logic around dealing with multiple .ngtypecheck files.

A new `TemplateTypeChecker` hosts that logic and interacts with the
`TypeCheckingProgramStrategy` to actually generate and return diagnostics.
The `TypeCheckContext` class, previously the workhorse of template type-
checking, is now solely focused on collecting and generating type-checking
file contents.

A side effect of implementing the new `TypeCheckingProgramStrategy` in this
way is that the API is designed to be suitable for use by the Angular
Language Service as well. The LS also needs to type-check components, but
has its own method for constructing a `ts.Program` with type-checking code.

Note that this commit does not make the actual checking of templates at all
_incremental_ just yet. That will happen in a future commit.

alxhub committed:
fix(compiler-cli): fix bug tracking indirect NgModule dependencies

The compiler needs to track the dependencies of a component, including any
NgModules which happen to be present in a component's scope. If an upstream
NgModule changes, any downstream components need to have their templates
re-compiled and re-typechecked.

Previously, the compiler handled this well for the A -> B -> C case where
module A imports module B which re-exports module C. However, it fell apart
in the A -> B -> C -> D case, because previously tracking focused on changes
to components/directives in the scope, and not NgModules specifically.

This commit introduces logic to track which NgModules contributed to a given
scope, and treat them as dependencies of any components within.

This logic also contains a bug, which is intentional for now. It
purposefully does not track transitive dependencies of the NgModules which
contribute to a scope. If it did, using the current dependency system, this
would treat all components and directives (even those not exported into the
scope) as dependencies, causing a major performance bottleneck. Only those
dependencies which contributed to the module's export scope should be
considered, but the current system is incapable of making this distinction.
This will be fixed at a later date.

alxhub committed:
perf(compiler-cli): perform template type-checking incrementally

This optimization builds on a lot of prior work to finally make type-
checking of templates incremental.

Incrementality requires two main components:
- the ability to reuse work from a prior compilation.
- the ability to know when changes in the current program invalidate that
  prior work.

Prior to this commit, on every type-checking pass the compiler would
generate new .ngtypecheck files for each original input file in the program.

1. (Build #1 main program): empty .ngtypecheck files generated for each
   original input file.

2. (Build #1 type-check program): .ngtypecheck contents overridden for those
   which have corresponding components that need type-checked.

3. (Build #2 main program): throw away old .ngtypecheck files and generate
   new empty ones.

4. (Build #2 type-check program): same as step 2.

With this commit, the `IncrementalDriver` now tracks template type-checking
_metadata_ for each input file. The metadata contains information about
source mappings for generated type-checking code, as well as some
diagnostics which were discovered at type-check analysis time. The actual
type-checking code is stored in the TypeScript AST for type-checking files,
which is now re-used between programs as follows:

1. (Build #1 main program): empty .ngtypecheck files generated for each
   original input file.

2. (Build #1 type-check program): .ngtypecheck contents overridden for those
   which have corresponding components that need type-checked, and the
   metadata registered in the `IncrementalDriver`.

3. (Build #2 main program): The `TypeCheckShimGenerator` now reuses _all_
   .ngtypecheck `ts.SourceFile` shims from build #1's type-check program in
   the construction of build #2's main program. Some of the contents of
   these files might be stale (if a component's template changed, for
   example), but wholesale reuse here prevents unnecessary changes in the
   contents of the program at this point and makes TypeScript's job a lot
   easier.

4. (Build #2 type-check program): For those input files which have not
   "logically changed" (meaning components within are semantically the same
   as they were before), the compiler will re-use the type-check file
   metadata from build #1, and _not_ generate a new .ngtypecheck shim.
   For components which have logically changed or where the previous
   .ngtypecheck contents cannot otherwise be reused, code generation happens
   as before.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -49,13 +52,28 @@ export class TemplateTypeChecker {
       this.files.set(file, {...fileData});
     }
+
+    this.priorBuild.recordSuccessfulTypeCheck(results.perFileData);
```
This call is unfortunate, reporting the type check results to the prior build. I understand this has been done to avoid a circular dependency between `IncrementalDriver` in `incremental` and `typecheck`, as they can both only see `incremental:api` (and therefore the `IncrementalBuild` interface).

Either:
1. Rename `priorBuild` to `incrementalBuild`, that would make this read somewhat more naturel.
2. Return `results.perFileData` here such that the caller `NgCompiler` can report the results to `IncrementalDriver`. This allows to `recordSuccessfulTypeCheck` to be dropped from the `IncrementalBuild` interface.

alxhub replied:
Ahh, very nice idea. Done.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -128,6 +128,22 @@ export class TypeCheckContext {
   private typeCtorPending = new Set<ts.ClassDeclaration>();
 
+  /**
+   * Map of data for file paths which was adopted from a prior compilation.
```
Given that it's not the map itself but its entries that are adopted, this should be:

```suggestion
   * Map of data for file paths which were adopted from a prior compilation.
```

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/shim.ts:
```diff
@@ -22,7 +22,16 @@ export class TypeCheckShimGenerator implements PerFileShimGenerator {
+      sf: ts.SourceFile, genFilePath: AbsoluteFsPath,
+      priorShimSf: ts.SourceFile|null): ts.SourceFile {
+    if (priorShimSf !== null) {
+      // If this shim existed in the previous program, reuse it now. It might not be correct, but
+      // reuse at this point will stop TypeScript doing unnecessary work, since later during the
+      // type-checking phase it's going to either be reused, or replaced. Thus there's no harm in
+      // reuse here even if it's out of date.
```
I would include somewhere in this comment that reusing the prior shim avoids changing the ES module imports compared to the prior build, enabling full program reuse (which is currently mentioned as "avoiding unnecessary work" without specifying what that work would be and why it is unnecessary)

JoostK commented on packages/compiler-cli/test/ngtsc/incremental_spec.ts:
```diff
@@ -70,6 +70,10 @@ runInEachFileSystem(() => {
     });
 
     it('should rebuild components that have changed', () => {
+      env.tsconfig({
+        ivyTemplateTypeCheck: true,
+        strictTemplates: true,
+      });
```
`ivyTemplateTypeCheck` shouldn't be relevant here.

```suggestion
      env.tsconfig({strictTemplates: true});
```

JoostK commented on packages/compiler-cli/test/ngtsc/incremental_spec.ts:
```diff
@@ -471,6 +475,261 @@ runInEachFileSystem(() => {
+      describe('inline operations', () => {
+        it('should still pick up on errors from inlined type check blocks', () => {
+          // In certain cases the template type-checker has to inline type checking blocks into user
+          // code, instead of placing it in a parallel template type-checking files. In these cases
```
```suggestion
          // code, instead of placing it in a parallel template type-checking file. In these cases
```

JoostK commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -480,17 +478,36 @@ export class NgCompiler {
+  }
+
+  private getTemplateDiagnostics(): ReadonlyArray<ts.Diagnostic> {
+    const host = this.host;
```
This isn't used anymore.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/adapter.ts:
```diff
@@ -0,0 +1,199 @@
+      oldProgram: ts.Program|null) {
+    // Initialize `this.generators` with a regex that matches each generator's paths.
+    for (const gen of perFileGenerators) {
+      const pattern = `^(.*)\\.${gen.extensionPrefix}\\.ts?$`;
```
👍 nice catch - I think I copy-pasted the regex and deleted the `x`. And good idea.

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -5097,6 +5097,7 @@ runInEachFileSystem(os => {
           const fileoverview = `
         /**
          * @fileoverview added by tsickle
+         * Generated from: test.ts
```
This threw me for a loop too when I saw it, actually. Maybe I should've pulled the change out into a separate commit, but this is the result of setting `rootDir` in our `tsconfig.json` (which needed to be set now that we weren't adding a `/__ng_typecheck__.ts` file at the top level).

If `rootDir` is present `tsickle` adds this comment: https://github.com/angular/tsickle/blob/1751f0b05fb6a5816c68a2b8ceda28ddcf6b94d1/src/fileoverview_comment_transformer.ts#L37-L41

JoostK commented:
One more thing: let `TraitCompiler` implement the `ProgramTypeCheckAdapter` interface (which requests it to be exported from the "typecheck" unit.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -0,0 +1,109 @@
+    const results = ctx.finalize();
+    this.typeCheckingStrategy.updateFiles(results.updates, UpdateMode.Complete);
+    for (const [file, fileData] of results.perFileData) {
+      this.files.set(file, {...fileData});
```
Does `fileData` really have to be shallow cloned? Where would it be mutated?

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/adapter.ts:
```diff
@@ -0,0 +1,199 @@
+      const inputFile = this.delegate.getSourceFile(baseFileName, ts.ScriptTarget.Latest);
+      if (inputFile === undefined || isShim(inputFile)) {
+        // Something strange happened here.
+        return null;
```
This situation isn't likely to occur, no. It's really only present so the typings work out.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/expando.ts:
````diff
@@ -0,0 +1,103 @@
+ * // sf is NgExtendedSourceFile
+ * ```
+ */
+export function extendSf(sf: ts.SourceFile): asserts sf is NgExtendedSourceFile {
````
Good idea, done.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/adapter.ts:
```diff
@@ -0,0 +1,199 @@
+  private generateSpecific(
+      fileName: AbsoluteFsPath, generator: PerFileShimGenerator,
+      inputFile: ts.SourceFile): ts.SourceFile {
+    const priorShimSf = this.priorShims.has(fileName) ? this.priorShims.get(fileName)! : null;
```
Good idea, done.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/reference_tagger.ts:
```diff
@@ -0,0 +1,74 @@
+
+import {extendSf, isExtended as isExtendedSf, isShim, NgExtension} from './expando';
+
+const TS_EXTENSIONS = /\.tsx?$/i;
```
I'm inclined to leave it - I feel like case insensitivity is more correct and we should be striving for it.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/reference_tagger.ts:
```diff
@@ -0,0 +1,74 @@
+ * Manipulates the `referencedFiles` property of `ts.SourceFile`s to add references to shim files
+ * for each original source file, causing the shims to be loaded into the program as well.
+ */
+export class ShimReferenceTagger {
```
😀

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/reference_tagger.ts:
```diff
@@ -0,0 +1,74 @@
+   *
+   * This is used to avoid generating shims twice for the same file.
+   */
+  private tagged = new Set<ts.SourceFile>();
```
Good point. I think the right thing to do is have an `enabled` mode, and disable that after cleanup.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/test/reference_tagger_spec.ts:
```diff
@@ -0,0 +1,83 @@
+import {TestShimGenerator} from './util';
+
+runInEachFileSystem(() => {
+  describe('ShimReferenceTagger', () => {
```
Good call! I added a test case specifically for this point.

alxhub commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -1838,6 +1838,29 @@ export declare class AnimationEvent {
          });
     });
+
+    describe('stability', () => {
```
Added to `ngtsc_spec` - good idea!

alxhub commented on packages/compiler-cli/src/ngtsc/core/src/compiler.ts:
```diff
@@ -380,28 +380,26 @@ export class NgCompiler {
-    const compilation = this.ensureAnalyzed();
-    // Run template type-checking.
+  private getTypeCheckingConfig(): TypeCheckingConfig {
+    // Determine the strictness level of type checking based on compiler options. As
```
Hmm. It seems fine to me - this is calculating the `TypeCheckingConfig`, which is determining how strict template type-checking is going to be.

alxhub commented on packages/compiler-cli/src/ngtsc/shims/src/adapter.ts:
```diff
@@ -164,7 +170,10 @@ export class ShimAdapter {
+        //
+        // Note: this path doesn't add to notShims because that would result in a `null` being
+        // returned in future requests.
+        return undefined;
```
I think I updated it as part of addressing feedback earlier - let me know if you feel like it's still not accurate.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -174,73 +243,71 @@ export class TypeCheckContext {
+  finalize(): TypeCheckRequest {
+    // First, build the map of updates to source files.
+    const updates = new Map<AbsoluteFsPath, string>();
+    for (const originalSf of this.program.getSourceFiles()) {
```
Maybe, but that would require looking up a `ts.SourceFile` from each operation. Also, there can be many operations for each `ts.SourceFile`, so it would require tracking uniqueness in some way.

JoostK replied:
My thought was to iterate just the keys, as those are the source files for which we have ops pending. This is how  `this.transform` currently fetches the ops for a source file, so its sets of keys is exactly the set of source files that need to be transformed.

alxhub replied:
Ahhh, gotcha. That makes total sense now. I think I looked at the signature of `adoptedFiles` and thought it was `opMap`.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_constructor_spec.ts:
```diff
@@ -74,12 +78,12 @@ TestClass.ngTypeCtor({value: 'test'});
         ]);
-        const ctx =
-            new TypeCheckContext(ALL_ENABLED_CONFIG, emitter, reflectionHost, _('/_typecheck_.ts'));
+        const ctx = new TypeCheckContext(ALL_ENABLED_CONFIG, program, emitter, reflectionHost);
```
Done! Good call.

alxhub commented on packages/compiler-cli/test/ngtsc/incremental_spec.ts:
```diff
@@ -251,6 +251,84 @@ runInEachFileSystem(() => {
+        import {NgModule} from '@angular/core';
+
+        @NgModule({})
+        export class MiddleBModule {}
```
It isn't - per the comment at the beginning, `MiddleBModule` is modified in the test to export `DirModule`, which changes the scope of the original module.

JoostK replied:
Good point, I hadn't considered the update.

JoostK commented on packages/compiler-cli/src/ngtsc/shims/src/expando.ts:
````diff
@@ -0,0 +1,103 @@
+
+/**
+ * Returns the `NgExtensionData` for a given `ts.SourceFile`, adding it if none exists.
+ * ```
````
```suggestion

```

JoostK commented on packages/compiler-cli/ngcc/src/analysis/ngcc_trait_compiler.ts:
```diff
@@ -82,8 +82,14 @@ export class NgccTraitCompiler extends TraitCompiler {
+    return null;
+  }
+
+  recordSuccessfulTypeCheck(): void {}
```
This can be removed

JoostK reviewed (approved):
Exciting! 💯

alxhub committed:
fixup! fix(compiler): switch to 'referencedFiles' for shim generation

This fixup adds per-file shims to root files when possible.

JoostK commented on packages/compiler-cli/src/ngtsc/core/test/host_test.ts:
```diff
@@ -0,0 +1,52 @@
+/**
```
This file would ideally be called `host_spec.ts`

JoostK commented on packages/compiler-cli/src/ngtsc/core/test/host_test.ts:
```diff
@@ -0,0 +1,52 @@
+        flatModuleOutFile: './entry',
+        generateNgFactoryShims: true,
+      };
+      const baseHost = new NgtscCompilerHost(getFileSystem(), options);
```
nit:

```suggestion
      const baseHost = new NgtscCompilerHost(fs, options);
```

alxhub committed:
fixup! fix(compiler): switch to 'referencedFiles' for shim generation

This fixes another g3 issue with importing ngfactory files from previous
compilations.

kara reviewed (approved):
LGTM for circular-dep change

Reviewed-for: circular-dependencies

filipesilva reviewed (approved):
Angular CLI will need changes to support some of these new features.

There's two main areas in which this PR affects CLI:
1. it introduces new files (the `.ngtypecheck.ts` ones) that don't seem to always be used
2. it may change the way errors are reported on rebuilds, which CLI tests

You can see the effects of 1. in the [AIO local build](https://circleci.com/gh/angular/angular/696887):
```
WARNING in /home/circleci/ng/aio/src/main.ngtypecheck.ts is part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your tsconfig.

WARNING in /home/circleci/ng/aio/src/app/app.module.ngtypecheck.ts is part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your tsconfig.

WARNING in /home/circleci/ng/aio/src/app/app.component.ngtypecheck.ts is part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your tsconfig.

WARNING in /home/circleci/ng/aio/src/app/documents/document.service.ngtypecheck.ts is part of the TypeScript compilation but it's unused.
Add only entry points to the 'files' or 'include' properties in your tsconfig.

...
```

Angular CLI issues these warnings to help users detect TS files they include but don't use in the compilation. I'm not sure if the use detection is failing here, if Ivy is creating extra references that aren't needed, or if CLI should just ignore these. 

[This](https://github.com/angular/angular-cli/blob/d008cf2b83f206f6223b68b014aba5098e8992c2/packages/ngtools/webpack/src/angular_compiler_plugin.ts#L624-L691) is the code that tries to identify the unused files. It incorporates some assumptions to effectively compute the list. Maybe these don't hold true with the way the extra files are referenced, but since we already had to ignore the typecheck files before I guess the previous mechanism didn't either. But it is sounds like the detection should work since `fix(compiler): switch to 'referencedFiles' for shim generation` says it uses the `ts.SourceFile.referencedFiles` facility.

But whatever the case, CLI users will see a lot of these warnings when using this updated compiler. 

In regards to 2., CLI would have to actually run this code through its test battery to identify changes in error reporting, if any.

This change seems to be master-only so I don't think we need to coordinate before it is merged. CLI changes can be done after this goes in, but still before 10.

cc @clydin @alan-agius4
